libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java

```diff
@@ -0,0 +1,146 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package androidx.media3.exoplayer.rtsp.reader;
+
+import static androidx.media3.common.util.Assertions.checkArgument;
+import static androidx.media3.common.util.Assertions.checkStateNotNull;
+
+import androidx.media3.common.C;
+import androidx.media3.common.Format;
+import androidx.media3.common.util.Log;
+import androidx.media3.common.util.ParsableByteArray;
+import androidx.media3.common.util.Util;
+import androidx.media3.exoplayer.rtsp.RtpPacket;
+import androidx.media3.exoplayer.rtsp.RtpPayloadFormat;
+import androidx.media3.extractor.ExtractorOutput;
+import androidx.media3.extractor.OpusUtil;
+import androidx.media3.extractor.TrackOutput;
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+
+/**
+ * Parses an OPUS byte stream carried on RTP packets and extracts individual samples. Refer to
+ * RFC7845 for more details.
+ */
+/* package */ final class RtpOpusReader implements RtpPayloadReader {
+  private static final String TAG = "RtpOpusReader";
+
+  private final RtpPayloadFormat payloadFormat;
+  private static final long MEDIA_CLOCK_FREQUENCY = 48_000;
+
+  private @MonotonicNonNull TrackOutput trackOutput;
+  private long firstReceivedTimestamp;
+  private long startTimeOffsetUs;
+  private int previousSequenceNumber;
+  private boolean foundOpusIDHeader;
+  private boolean foundOpusCommentHeader;
+
+  public RtpOpusReader(RtpPayloadFormat payloadFormat) {
+    this.payloadFormat = payloadFormat;
+    this.firstReceivedTimestamp = C.INDEX_UNSET;
+    this.previousSequenceNumber = C.INDEX_UNSET;
+    this.foundOpusIDHeader = false;
+    this.foundOpusCommentHeader = false;
+  }
+
+  // RtpPayloadReader implementation.
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, int trackId) {
+    trackOutput = extractorOutput.track(trackId, C.TRACK_TYPE_AUDIO);
+    trackOutput.format(payloadFormat.format);
+  }
+
+  @Override
+  public void onReceivingFirstPacket(long timestamp, int sequenceNumber) {
+    this.firstReceivedTimestamp = timestamp;
+  }
+
+  @Override
+  public void consume(
+      ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker) {
+    checkStateNotNull(trackOutput);
+
+    /* RFC7845 Section 3.
+     * +---------+ +----------------+ +--------------------+ +-----
+     * |ID Header| | Comment Header | |Audio Data Packet 1 | | ...
+     * +---------+ +----------------+ +--------------------+ +-----
+     */
+    if (!foundOpusIDHeader) {
```

Comment:
Sorry for not getting back to this sooner, I thought we have merged it.

All the logic here to find Opus ID and comment headers imply that the header and comment header is ONE PACKET long. But that is not what is specified in the RFC.

According to the RFC, the ID header is one PAGE long, and the comment header can span across many pages. And one page can span multiple RTP packets. Also, "Audio data packets might span page boundaries.  The first audio data page could have the 'continued packet' flag set"

Although this is working right now, it's not semantically correct. So please try to make it compliant with the RFC by supporting multiple packet pages. There are examples of one access unit spans multiple RTP packets, like in the H264/265 reader. It's probably hard to test, but given you have some tests going on, please add some test based on your interpretation to the RFC.